### DISCUSSION TITLE
Various Ideas: 'Meta Detection' Wiring Components

### DISCUSSION THREAD

**Discussion by Ek-Videogames** [FakeFishGames/Barotrauma#7595] [Ideas and suggestions]
There are certain situations in wiring systems that submakers cannot account for such as the 'renegade' sub systems being mirrored but their angles being the same, or the performance/effects of machines being altered by upgrades applied to them.

The general idea i'm suggesting is some kind of 'meta' signal component or multiple-components. 

This is just one example of possible implementation but: what if there was a set of public 'metadata tags' much like how variables exist in code, which can be read and output by these 'meta' components. 
- So if you wanted to figure out if the submarine was being mirrored because of a PVP mission / pirate mission there would be some property or set tag for "IsMirrored" and then the component would output if the sub was mirrored or not.
- Or, if the crush depth had been upgraded there would be some 'meta-tag' which reflected the level of the hull upgrade.
`( I'm thinking output upgrade plus/minus% instead of 'upgrade level' because that's more useful for math & would be more compatible with modded machines or upgrades )`
- This can be applied as far as even more abstract sense, like being able to use campaign data statistics. The sky is the limit.

**Comment by mLuby** [comment]
Mirroring should be easy enough to detect by comparing the two Nav.`current_position_x` pins. 

Upgrades are trickier, but some if not most can be detected without resorting to metadata. The second column from the left [should all be detectable upgrades](https://gist.github.com/mLuby/e9c36a2f7a858d886eaba0675f2ab75b) (pre-`condition_out` so there should be even more now), usually by having a tiny sealed off setup which the circuit interrogates (e.g. power consumption for the efficiency upgrades, o2/water levels for the o2gen/pump output upgrades).

Maybe there are other examples of things only a new component could detect, like reputation? Not sure why you'd want that though.

**Reply by ubertpendragon** [reply]
I have indeed used the `current_position_x` pins in one of my PvP subs to reverse the controls of a sonar railgun, so it is possible
